include/zephyr/net/prometheus/gauge.h

```diff
@@ -27,39 +28,48 @@
  */
 struct prometheus_gauge {
 	/** Base of the Prometheus gauge metric */
-	struct prometheus_metric *base;
+	struct prometheus_metric base;
 	/** Value of the Prometheus gauge metric */
 	double value;
+	/** User data */
+	void *user_data;
 };
 
 /**
  * @brief Prometheus Gauge definition.
  *
- * This macro defines a Gauge metric.
+ * This macro defines a Gauge metric. If you want to make the gauge static,
+ * then add "static" keyword before the PROMETHEUS_GAUGE_DEFINE.
  *
- * @param _name The channel's name.
- * @param _detail The metric base.
+ * @param _name The gauge metric name.
+ * @param _desc Gauge description
+ * @param _label Label for the metric. Additional labels can be added at runtime.
+ * @param _collector Collector to map this metric. Can be set to NULL if it not yet known.
+ * @param ... Optional user data specific to this metric instance.
  *
  * Example usage:
  * @code{.c}
  *
- * struct prometheus_metric http_request_gauge = {
- *	.type = PROMETHEUS_GAUGE,
- *	.name = "http_request_gauge",
- *	.description = "HTTP request gauge",
- *	.num_labels = 1,
- *	.labels = {
- *		{ .key = "http_request", .value = "request_count",}
- *	},
- * };
- *
- * PROMETHEUS_GAUGE_DEFINE(test_gauge, &test_gauge_metric);
+ * PROMETHEUS_GAUGE_DEFINE(http_request_gauge, "HTTP request gauge",
+ *                         ({ .key = "http_request", .value = "request_count" }),
+ *                         NULL);
  *
  * @endcode
  */
-#define PROMETHEUS_GAUGE_DEFINE(_name, _detail)                                                    \
-	static STRUCT_SECTION_ITERABLE(prometheus_gauge, _name) = {.base = (void *)(_detail),      \
-								   .value = 0}
+#define PROMETHEUS_GAUGE_DEFINE(_name, _desc, _label, _collector, ...)	\
+	STRUCT_SECTION_ITERABLE(prometheus_gauge, _name) = {		\
+		.base.name = STRINGIFY(_name),				\
+		.base.type = PROMETHEUS_GAUGE,				\
+		.base.description = _desc,				\
+		.base.labels[0] = __DEBRACKET _label,			\
+		.base.num_labels = 1,					\
+		.base.collector = _collector,				\
+		.value = 0.0,						\
+		.user_data = COND_CODE_0(				\
+			NUM_VA_ARGS_LESS_1(LIST_DROP_EMPTY(__VA_ARGS__, _)), \
+			(NULL),						\
+			(GET_ARG_N(1, __VA_ARGS__))),			\
```

Comment:
Could this macro construction be turned into a helper macro that makes this more readable?

Reply:
Hmm, not sure. I copied this pattern from other part of the system, but I can try to simplify that in following PR.